Mark default parameters of operators as non-bound

OmarTawfik opened the pull request:
Fixes #14688 
@dotnet/roslyn-compiler 
@dotnet/roslyn-ide 
Tagging IDE because it is originally a crash in the ClassificationService.

jasonmalinowski commented on src/EditorFeatures/VisualBasicTest/Classification/SemanticClassifierTests.vb:
```diff
@@ -543,5 +543,21 @@ end sub
+End Class
+</code>.NormalizedValue()
+
+            Await TestAsync(text,
```
Shouldn't Foo be classified three times? Even if Optionals aren't allowed, I wouldn't expect it to fall down that badly.

CyrusNajmabadi replied:
Agreed.  Lots of stuff isn't legal in the language (for example, an optional parameter preceding a required parameter).  But we still get the experience right even in those cases.

OmarTawfik commented:
@jasonmalinowski the compiler (as called from the command line) behaves fine. This is a specific code path that only the classification service exhausts by iterating over tokens in the syntax tree and calling the binder on each of them. I'd rather test that this exact scenario works fine when called from the IDE layer.

CyrusNajmabadi commented on src/Compilers/VisualBasic/Portable/Binding/BinderFactory.vb:
```diff
@@ -387,8 +387,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                                ' Default values are not valid (and not bound) for lambda parameters or property accessors
+                                 SyntaxKind.RaiseEventAccessorStatement,
+                                 SyntaxKind.OperatorStatement
+                                ' Default values are not valid (and not bound) for these types of statements
```
I'm nto sure how i feel about that.  A primary purpose of hte Roslyn APIs is to give us semantic information, even for invalid code.

jasonmalinowski commented:
@OmarTawfik The Roslyn compiler is an API first and foremost. If exceptions are being thrown from APIs like this, then it's not working fine. I'm assuming we have tests somewhere that do the binding on all token checks in the compiler layer?

CyrusNajmabadi commented:
> I'd rather test that this exact scenario works fine when called from the IDE layer.

Note, the fact that it isn't classifying means that it's not "working fine" :)

OmarTawfik commented:
@CyrusNajmabadi @jasonmalinowski I believe we are talking about two issues here.

1. whether we should test this bug from the IDE layer, which I argue we should, as this code path is used in the classification service, not the compiler service.
2. whether the classification service is classifying all tokens or not, which I see now it isn't. Even though in the UI the keyword `Foo` is classified three times as a class (and showing in IDE in light blue), it returns in the test with two occurrences only. The missing one is the one in `Public Class Foo`. I'm looking now to see why not, suspecting it is a test issue.

CyrusNajmabadi commented:
> whether we should test this bug from the IDE layer, which I argue we should, as this code path is used in the classification service, not the compiler service.

The compiler exposes an API that can hit htis issue.  We should test that API.  It's also ok to test the specific IDE scenario.  But we should def test out the actual API.

> whether the classification service is classifying all tokens or not, which I see now it isn't. Even though in the UI the keyword Foo is classified three times as a class (and showing in IDE in light blue), it returns in the test with two occurrences only.

You can move over to using the "TotalClassifierTest" harness.  It sounds like you're just in the SemanticClassifier harness.

jasonmalinowski commented:
Ah, if it's missing the first one then that's probably because the first one is applied by the syntactic classifier not and the semantic one.

I still disagree on the first point. The bug is in the compiler. The compiler is an API that just happens to have an emit method. Just because you can't hit it through the command line compiler does not make it "not a compiler bug".

IDE tests might sometimes be appropriate if the IDE patterns were interesting. For example, if some crazy case in rename hit a compiler bug, I'd say we should test both the compiler directly and also the IDE. This doesn't meet the bar in my book here. I don't see any reason IDE code should be touched as a part of this pull request since there's nothing special the IDE is doing other than calling the compiler API.

OmarTawfik commented:
> IDE tests might sometimes be appropriate if the IDE patterns were interesting. For example, if some crazy case in rename hit a compiler bug, I'd say we should test both the compiler directly and also the IDE. This doesn't meet the bar in my book here. I don't see any reason IDE code should be touched as a part of this pull request since there's nothing special the IDE is doing other than calling the compiler API.

@jasonmalinowski @CyrusNajmabadi fixed the comments and added another test to the compilers layer.
I removed the test from the IDE layer, even though I don't see harm in testing both layers. LMK if you decided otherwise.

CyrusNajmabadi commented:
> I removed the test from the IDE layer, even though I don't see harm in testing both layers. LMK if you decided otherwise.

I'm totally fine with tests at both levels.  I think jason's overall point though was that this is best tested at the compiler layer as it's a compiler API problem.

CyrusNajmabadi commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/UserDefinedBinaryOperators.vb:
```diff
@@ -2546,6 +2546,35 @@ End Module
+
+            Dim model = GetSemanticModel(compilation, "a.vb")
+            Dim node As ExpressionSyntax = CompilationUtils.FindBindingText(Of IdentifierNameSyntax)(compilation, "a.vb", 1)
+            Assert.NotNull(model.GetSymbolInfo(node)) 'Compiler API can retreive information about symbol without exceptions
```
Probably good to validate at least that you get a good symbol back.  I.e. it's a NamedTypeSymbol, it's not an ErrorType, and the name is 'Foo'.

AlekseyTs replied:
Yes, please validate what symbol do you get back. Also, please test GetTypeInfo and GetAliasInfo APIs.

jaredpar commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/UserDefinedBinaryOperators.vb:
```diff
@@ -2546,6 +2546,35 @@ End Module
+<compilation name="OperatorsWithDefaultValuesAreNotBound">
+    <file name="a.vb"><![CDATA[
+Public Class Foo
+    Public Shared Operator Not(Optional x As Foo = Nothing) As Integer 'BIND1:"Foo"
```
Make sure there is a corresponding test where there is not a default value:

``` vb
Public Shared Operator Not(Optional x As Foo) As Integer 'BIND1:"Foo"
```

AlekseyTs replied:
+1

AlekseyTs replied:
>Nothing [](start = 51, length = 7)

Please add a test that uses some meaningful default value and verifies SemanticModel behavior within the value.

jaredpar commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/UserDefinedBinaryOperators.vb:
```diff
@@ -2546,6 +2546,35 @@ End Module
 
         End Sub
 
+
```
Nit: double blank line.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/BinderFactory.vb:
```diff
@@ -387,8 +387,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                                 SyntaxKind.RaiseEventAccessorStatement
-                                ' Default values are not valid (and not bound) for lambda parameters or property accessors
+                                 SyntaxKind.RaiseEventAccessorStatement,
+                                 SyntaxKind.OperatorStatement
```
>SyntaxKind.OperatorStatement [](start = 33, length = 28)

It is not clear why are we getting here for the test scenario. We are not asking questions about default value. It looks like ```Public Overrides Function VisitParameter(node As ParameterSyntax) As Binder``` should be changed changed to request a binder for NodeUsage.ParameterDefaultValue only if we are after non-missing ```=``` token.

OmarTawfik replied:
@AlekseyTs I'm not sure I'm following. The fix here is to change `CreateBinderForNodeAndUsage()` to return `Nothing` if the caller asks for a `NodeUsage.ParameterDefaultValue` with a `SyntaxKind.OperatorStatement`. Callers of `GetBinderForNodeAndUsage()` should expect and know how to handle return values of `Nothing`. Why should the upper level `VisitParameter()` check for specific node types if the underlying `BinderFactory` type is designed to do so? Am I missing something?

OmarTawfik replied:
@AlekseyTs checked the call stack. Basically `ParseExpressionCore(bailIfFirstTokenRejected)` is false when parsing optional parameters, so it returns an empty expression instead of `Nothing`. In all checks through the code base, we check whether the default value of the parameter is `Nothing` or not. I'm changing it now, but is there any value of using an empty expression syntax here?

jaredpar commented:
Ping. This PR seems to have gone stale.  What do we need to get this moving again?

OmarTawfik commented:
@jaredpar will open a new PR with a better proposed fix/rebased on latest from master.
